### ISSUE TITLE
Hot Restart hangs on Stable 2.0.2

### ISSUE THREAD

**Issue by crivatz11** [flutter/flutter#78570]
I'm running the latest Stable Flutter version on my Mac OS Big sur, Macbook pro M1, testing a web app.
I've worked on a couple of apps the past few weeks and everything worked fine.

Suddenly, while in the middle of development on a particular app, the Hot Reload feature stopped working properly.
I save some changes to a file, it starts, and keeps loading forever. The only way to see my changes is to Stop the app and Run it again.

Here is the output for 'flutter doctor -v':

[✓] Flutter (Channel stable, 2.0.2, on macOS 11.2.3 20D91 darwin-arm, locale en-RO)
    • Flutter version 2.0.2 at /Users/crivatz/Documents/Dev/flutter
    • Framework revision 8962f6dc68 (7 days ago), 2021-03-11 13:22:20 -0800
    • Engine revision 5d8bf811b3
    • Dart version 2.12.1

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).

[✓] VS Code (version 1.54.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.20.0

[✓] Connected device (1 available)
    • Chrome (web) • chrome • web-javascript • Google Chrome 89.0.4389.90

! Doctor found issues in 2 categories.

**Comment by darshankawar** [comment]
@crivatz11 
Does it happen in debug mode or in release mode ? After making changes to a file and pressing `r`, does it just hang at `performing restart` or comes back after waiting for few seconds (ex: 20 or 30 seconds) ?

There's similar open issue which reported `hot restart` taking a bit time after making changes to files, https://github.com/flutter/flutter/issues/78239

**Comment by crivatz11** [comment]
It happens on debug mode, I've waited for a few minutes, nothing happens.
On release mode (which I haven't tried until you suggested) it is actually successful in restarting (it takes about 20 secs).

However when in press "Debug" in release mode, the VSCode extension keeps saying "Compiling lib/main.dart for the Web...", even after it loads the app, restarts it a few times... looks like it's going to keep saying it forever.
Not sure if it's related, could be a new issue.

**Comment by darshankawar** [comment]
@crivatz11 
I don't have VSCode setup on my machine since I use AS.  Tried a sample app using latest stable (2.0.3) and was able to hot restart in debug mode without any issues.

<img width="679" alt="Screenshot 2021-03-22 at 11 55 21 AM" src="https://user-images.githubusercontent.com/67046386/111948752-a9838c00-8b05-11eb-90fc-cb33ecacef16.png">

I am using CatalinaOS, but not sure if it's specific to BigSur, M1 + VScode combination issue.

Can you provide us a [complete minimal reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example
) along with steps to replicate that clearly shows the issue you are facing ?


<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.0.3, on Mac OS X 10.15.4 19E2269 darwin-x64,
    locale en-GB)
    • Flutter version 2.0.3 at /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision 4d7946a68d (3 days ago), 2021-03-18 17:24:33 -0700
    • Engine revision 3459eb2436
    • Dart version 2.12.2

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin
        code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see
      https://guides.cocoapods.org/using/getting-started.html#installation for
      instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.54.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.1

[✓] Connected device (3 available)
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                               •
      web-javascript • Google Chrome 89.0.4389.90

! Doctor found issues in 1 category.

```

</details>

**Comment by nmfisher** [comment]
I've noticed hot reload is a bit flaky after upgrading master, too (hanging after the third or fourth hot reload). This only seems to happen on my main app with an extra isolate running in the background, though, not smaller apps with a single isolate. I'll see if I can reproduce.

[+3854 ms] [√] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
[   +1 ms]     • Android SDK at C:\Users\Nick\AppData\Local\Android\sdk
[   +1 ms]     • Platform android-30, build-tools 29.0.3
[   +2 ms]     • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
[   +1 ms]     • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b04)
[        ]     • All Android licenses accepted.
[   +8 ms] [√] Chrome - develop for the web
[   +6 ms]     • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe
[  +11 ms] [√] Android Studio (version 3.6)
[        ]     • Android Studio at C:\Program Files\Android\Android Studio
[  +10 ms]     • Flutter plugin can be installed from:
[   +9 ms]        https://plugins.jetbrains.com/plugin/9212-flutter
[  +10 ms]     • Dart plugin can be installed from:
[        ]        https://plugins.jetbrains.com/plugin/6351-dart
[  +10 ms]     • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b04)
[  +15 ms] [√] VS Code (version 1.54.3)
[   +7 ms]     • VS Code at C:\Users\Nick\AppData\Local\Programs\Microsoft VS Code
[  +13 ms]     • Flutter extension version 3.20.0
[  +15 ms] [√] Connected device (3 available)
[   +7 ms]     • SM J260FU (mobile) • aca90691 • android-arm    • Android 8.1.0 (API 27)
[   +6 ms]     • Chrome (web)       • chrome   • web-javascript • Google Chrome 89.0.4389.90
[   +5 ms]     • Edge (web)         • edge     • web-javascript • Microsoft Edge 89.0.774.57

**Comment by crivatz11** [comment]
@darshankawar 
For me, it happens when debugging on Web. I didn't run the app on mobile devices yet.
I tried to pinpoint the exact problem and I ran into something very strange:

1. I found that it only happens on a particular widget on my app (representing a whole page). For all other pages/widgets, hot reload works fine.
2. I tried cutting code from the problematic widget, **until I simply had a basic Stateless Widget with a Text() child** and Hot Reload still didn't work. But again, for other widgets in other files, it worked fine.
3. The only way I am able to use Hot Reload on the problematic widget/page/file is to let Hot Reload start and then manually hit refresh on my Chrome Debugger instance. After pressing refresh, I get a success message and changes are shown correctly.

**Comment by crivatz11** [comment]
@darshankawar 
So for a particular .dart file, believe it or not, the following code doesn't hot reload unless I also manually hit refresh on the browser:
`import 'package:flutter/material.dart';

class NewUserGeneralData extends StatelessWidget {
  const NewUserGeneralData({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      child: Text('Test'),
    );
  }
}`

**Comment by darshankawar** [comment]
@crivatz11 
Tried your code sample using latest stable on web (`flutter run -d chrome`) and making changes to the `Text` widget and hot restarting with `r`, did update on screen without any issues.

```
Performing hot restart...                                          150ms
Restarted application in 150ms.

Performing hot restart...                                          160ms
Restarted application in 160ms.

Performing hot restart...                                          156ms
Restarted application in 156ms.

```

```
return Container(
      child: Text('Test test test Test'),
    );
```

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.0.3, on Mac OS X 10.15.4 19E2269 darwin-x64,
    locale en-GB)
    • Flutter version 2.0.3 at /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision 4d7946a68d (5 days ago), 2021-03-18 17:24:33 -0700
    • Engine revision 3459eb2436
    • Dart version 2.12.2

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin
        code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see
      https://guides.cocoapods.org/using/getting-started.html#installation for
      instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.54.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.20.0

[✓] Connected device (3 available)
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                               •
      web-javascript • Google Chrome 89.0.4389.90

! Doctor found issues in 1 category.

```

</details>

If this doesn't represent your issue, please provide more details as to exactly replicate it. A video / gif would be helpful too.

**Comment by darshankawar** [comment]
@nmfisher I'll comeback to your issue.

**Comment by crivatz11** [comment]
Hi @darshankawar 
I sent you an invite to my repo so you can check out the code or clone it and test it out if you prefer.
Not sure it this is the correct way, let me know otherwise.

It seems to be related to any page in the "new_user_pages" folder.
For the other pages, hot restart seems to work fine...

**Comment by darshankawar** [comment]
@crivatz11 
Thanks, but it'll be good if you could provide a [minimal reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example) that involves `new_user_pages` folder case, that we can directly copy paste and run to verify the issue.

**Comment by nmfisher** [comment]
@darshankawar just wondering if you'd looked into the isolate issue, or if I should open an other issue for that? Hot reload is basically unusable for me at the moment.

**Comment by darshankawar** [comment]
@nmfisher 
I looked at your code sample and noticed you've used `flutter_isolate` plugin which is 3rd party. Does issue persist if you remove that plugin and it's relevant code ? If not, then it may be plugin specific issue and you may need to file it in dedicated repo: https://github.com/rmawatson/flutter_isolate/issues.
If issue still persists, please provide updated code sample that doesn't contain the said plugin's usage.
Thanks.

**Comment by nmfisher** [comment]
@darshankawar removing flutter_isolate (and replacing with a vanilla Dart isolate) does seem to fix the problem. However, this isn't really a solution, because standard Dart isolates break two-way plugin channels (https://github.com/flutter/flutter/issues/13937).

If a plugin can break Flutter/Dart hot reload, doesn't that mean hot reload is broken?

**Comment by crivatz11** [comment]
@darshankawar 
Here it is:
[iscreen_flutter-master.zip](https://github.com/flutter/flutter/files/6229102/iscreen_flutter-master.zip)

All you need to do is open the side menu and click on "Inscrie-te".
You should then find yourself in the first page from the "new_user_pages" folder which doesn't hot restart properly.

**Comment by darshankawar** [comment]
@crivatz11 
Thanks for reproducible code sample. I tried it on latest stable by opening the menu from drawer and then hot restarting with `r` which restarted the screen properly. Also changed minor text and hot restarted which reflected properly without any issues, as below:


https://user-images.githubusercontent.com/67046386/113096974-ce69b480-9213-11eb-84e8-e2e842157895.mov


<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.0.3, on Mac OS X 10.15.4 19E2269 darwin-x64,
    locale en-GB)
    • Flutter version 2.0.3 at /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision 4d7946a68d (12 days ago), 2021-03-18 17:24:33 -0700
    • Engine revision 3459eb2436
    • Dart version 2.12.2

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin
        code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see
      https://guides.cocoapods.org/using/getting-started.html#installation for
      instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.54.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.20.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • Mac OS X 10.15.4 19E2269
      darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 89.0.4389.114

! Doctor found issues in 1 category.

```

</details>

**Comment by nmfisher** [comment]
@darshankawar still not sure if I was experiencing the same issue as OP, but this seems to be fixed on 2.1.0-12.2.pre (possibly earlier versions too, I just skipped to beta for other reasons).

**Comment by pedromassangocode** [comment]
Without additional information, we are unfortunately not sure how to resolve this issue. 
We are therefore reluctantly going to close this bug for now. 
Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!  
Thanks for your contribution.

Could everyone who still has this problem please file a new issue with the exact description of what happens, logs, and the output of `flutter doctor -v`.
All system setups can be slightly different, so it's always better to open new issues and reference related issues.
